### ISSUE TITLE
Save additional load data in Frame

### ISSUE THREAD

**Issue by asajeffrey** [servo/servo#14888]
After #14312, documents may be discarded and reloaded from the information stored in a `Frame` object. Currently we are just storing the URL, we should really also be storing the origin as well, so that when the document is reloaded, we can restore its origin. cc @jdm @cbrewster

**Comment by bzbarsky** [comment]
You probably need to store other things too.  What Gecko stores (in nsSHEntry) is at least:

1. URI.
2. "original URI" (may not be relevant to servo; this is for cases when the URI of the document doesn't match the URI that needs to actually be loaded to load the document, due to internal redirects).
3. Referrer that was used when doing the original load (on the Gecko end).
4. Referrer policy to use, so the actual referrer to send can be computed properly.
5. POST data, if any.  Also cache keys to allow redoing the fetch from cache.
6. Viewport scroll position.
7. State data (from pushstate/popstate).  Not sure whether this is really used in the post-discard case or just for history traversals across push/pop.
8. Srcdoc data, if any.
9. Base URI (needed for the srcdoc case).
10. Origins (triggering and loading) used during the original fetch, so the resulting origin can be recomputed properly, security policies can be enforced properly, etc.  This kind of covers the "origin of the page", since that is computed based on those and the URL.
11. A MIME type hint.  This is needed to support `<a type=foo>` properly.
12. State of form controls, for form state restoration.
13. State of pending meta refreshes.

**Comment by asajeffrey** [comment]
IRC chat with @bzbarsky: http://logs.glob.uno/?c=mozilla%23servo&s=6+Jan+2017&e=6+Jan+2017#c588746

TL;DR: we may be able to repurpose `LoadData` for this, if it stored the `origin` used for the case where the URL is `about:blank`.

**Comment by asajeffrey** [comment]
See #14892 for a wrinkle.

**Comment by cbrewster** [comment]
There is still more data to be added to `FrameState`

**Comment by nox** [comment]
@asajeffrey Are you still working on this?
@cbrewster What additional data?

**Comment by asajeffrey** [comment]
@nox: the additional data is what @bz mentioned. Much of this should be in `LoadData` rather than `SessionHistoryEntry`. This should be the same notion of load data as the embedding API uses (cc @paulrouget).

The one I don't understand is the form data, which requires some heuristics to map discarded-document form elements to reloaded-document form elements.

**Comment by paulrouget** [comment]
A few other things we might want to store:

- http_method (post or get)
- scroll_restoration_mode (auto / manual)
- transition_type (see https://github.com/servo/servo/issues/18211)

These will be necessary too, but can be saved at the embedder level:
- page title
- page favicon
- user_typed_value

**Comment by asajeffrey** [comment]
Hmm, that transition_type seems very specific to the embedder, I wonder if there's a way to make it more generic? Something like giving the embedder an id they can use to associate state to a transition if they like? The pipeline id is already almost good enough, but I'm not sure we're exposing it in a way that's useful to the embedder.
